packages/workspace/src/core/project-graph/project-graph.ts

```diff
@@ -38,30 +39,85 @@ import {
   buildWorkspaceProjectNodes,
 } from './build-nodes';
 
+export const NEXT_GRAPH_VERSION = '4.0';
+export const CURRENT_GRAPH_VERSION = '3.0';
+
 /**
  * Synchronously reads the latest cached copy of the workspace's ProjectGraph.
  * @throws {Error} if there is no cached ProjectGraph to read from
+
+ * @param {string} projectGraphVersion Version to map ProjectGraph to
+ * @returns {ProjectGraph}
  */
-export function readCachedProjectGraph(): ProjectGraph {
+export function readCachedProjectGraph(
+  projectGraphVersion?: string
+): ProjectGraph {
   const projectGraphCache: ProjectGraphCache | false = readCache();
   if (!projectGraphCache) {
     throw new Error(`
       [readCachedProjectGraph] ERROR: No cached ProjectGraph is available.
-      
+
       If you are leveraging \`readCachedProjectGraph()\` directly then you will need to refactor your usage to first ensure that
       the ProjectGraph is created by calling \`await createProjectGraphAsync()\` somewhere before attempting to read the data.
 
       If you encounter this error as part of running standard \`nx\` commands then please open an issue on https://github.com/nrwl/nx
     `);
   }
   return {
-    nodes: projectGraphCache.nodes,
+    version: projectGraphCache.version,
+    nodes: projectNodesCompatAdapter(
+      projectGraphCache.nodes,
+      projectGraphVersion || projectGraphCache.version
+    ),
     dependencies: projectGraphCache.dependencies,
   };
 }
 
-export async function createProjectGraphAsync(): Promise<ProjectGraph> {
-  return createProjectGraph();
+/**
+ * Backwards compatibility adapter for project Nodes
+ * @param {Record<string, ProjectGraphNode>} nodes
+ * @param {string?} version
+ * @returns
+ */
+export function projectNodesCompatAdapter(
```

Comment:
Why isn't this invoked in createProjectGraph?

Reply:
For the time being it was not necessary as `readWorkspaceFiles` already uses `projectFileDataCompatAdapter`. But I can add it there already.